File: apps/api/src/common/controllers/health.controller.ts
import { Controller, Get } from '@nestjs/common';
import { ApiTags, ApiOperation } from '@nestjs/swagger';
import { DataSource } from 'typeorm';
import { SkipThrottle } from '@nestjs/throttler';

@ApiTags('Health')
@Controller('health')
@SkipThrottle()
export class HealthController {
  constructor(private readonly dataSource: DataSource) {}

  @Get()
  @ApiOperation({ summary: 'Health check' })
  async check() {
    const dbOk = this.dataSource.isInitialized;
    let dbLatency = -1;

    if (dbOk) {
      const start = Date.now();
      try {
        await this.dataSource.query('SELECT 1');
        dbLatency = Date.now() - start;
      } catch {
        dbLatency = -1;
      }
    }

    return {
      status: dbOk ? 'ok' : 'degraded',
      timestamp: new Date().toISOString(),
      uptime: Math.round(process.uptime()),
      database: {
        connected: dbOk,
        latency: dbLatency >= 0 ? `${dbLatency}ms` : 'error',
      },
      memory: {
        rss: `${Math.round(process.memoryUsage().rss / 1024 / 1024)}MB`,
        heap: `${Math.round(process.memoryUsage().heapUsed / 1024 / 1024)}MB`,
      },
    };
  }
}

File: apps/api-java/src/main/java/br/com/menufacil/controller/HealthController.java
package br.com.menufacil.controller;

import io.swagger.v3.oas.annotations.Operation;
import io.swagger.v3.oas.annotations.tags.Tag;
import lombok.RequiredArgsConstructor;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;

import javax.sql.DataSource;
import java.sql.Connection;
import java.time.LocalDateTime;
import java.util.LinkedHashMap;
import java.util.Map;

@Tag(name = "Health", description = "Status do sistema")
@RestController
@RequestMapping("/health")
@RequiredArgsConstructor
public class HealthController {

    private final DataSource dataSource;

    @Operation(summary = "Health check")
    @GetMapping
    public ResponseEntity<Map<String, Object>> health() {
        Map<String, Object> result = new LinkedHashMap<>();
        boolean dbOk = false;
        long dbLatency = -1;

        try (Connection conn = dataSource.getConnection()) {
            long start = System.currentTimeMillis();
            conn.createStatement().execute("SELECT 1");
            dbLatency = System.currentTimeMillis() - start;
            dbOk = true;
        } catch (Exception ignored) {}

        result.put("status", dbOk ? "ok" : "degraded");
        result.put("timestamp", LocalDateTime.now().toString());
        result.put("runtime", "spring-boot");
        result.put("database", Map.of(
                "connected", dbOk,
                "latency", dbLatency >= 0 ? dbLatency + "ms" : "error"
        ));
        result.put("memory", Map.of(
                "total", Runtime.getRuntime().totalMemory() / 1024 / 1024 + "MB",
                "free", Runtime.getRuntime().freeMemory() / 1024 / 1024 + "MB",
                "used", (Runtime.getRuntime().totalMemory() - Runtime.getRuntime().freeMemory()) / 1024 / 1024 + "MB"
        ));

        return ResponseEntity.ok(result);
    }
}
